### DISCUSSION TITLE
Failed to set migration status

### DISCUSSION THREAD

**Discussion by chrisonline** [kimai/kimai#4766] [On-Premise: Installation & Updates]
I have just updated the docker image, but after start I see in the logs following error at migrating to the new version:
```
Database `kimai` for connection named default already exists. Skipped.
[notice] Migrating up to DoctrineMigrations\Version20240326125247
[error] Migration DoctrineMigrations\Version20231130000719 failed during Execution. Error: "An exception occurred while executing a query: SQLSTATE[HY000]: General error: 1093 You can't specify target table 'u1' for update in FROM clause"
 [ERROR] Failed to set migration status: An exception occurred while executing a
         query: SQLSTATE[HY000]: General error: 1093 You can't specify target   
         table 'u1' for update in FROM clause                                   
+ /opt/kimai/bin/console -n kimai:update
```

After login I see this error in the logs:

```

[2024-04-10T08:36:21.780713+00:00] request.CRITICAL: Uncaught PHP Exception Twig\Error\RuntimeError: "An exception has been thrown during the rendering of a template ("An exception occurred while executing a query: SQLSTATE[42S22]: Column not found: 1054 Unknown column 'k0_.number' in 'field list'")." at datatable.html.twig line 14 {"exception":"[object] (Twig\\Error\\RuntimeError(code: 0): An exception has been thrown during the rendering of a template (\"An exception occurred while executing a query: SQLSTATE[42S22]: Column not found: 1054 Unknown column 'k0_.number' in 'field list'\"). at /opt/kimai/templates/datatable.html.twig:14)\n[previous exception] [object] (Doctrine\\DBAL\\Exception\\InvalidFieldNameException(code: 1054): An exception occurred while executing a query: SQLSTATE[42S22]: Column not found: 1054 Unknown column 'k0_.number' in 'field list' at /opt/kimai/vendor/doctrine/dbal/src/Driver/API/MySQL/ExceptionConverter.php:67)\n[previous exception] [object] (Doctrine\\DBAL\\Driver\\PDO\\Exception(code: 1054): SQLSTATE[42S22]: Column not found: 1054 Unknown column 'k0_.number' in 'field list' at /opt/kimai/vendor/doctrine/dbal/src/Driver/PDO/Exception.php:28)\n[previous exception] [object] (PDOException(code: 42S22): SQLSTATE[42S22]: Column not found: 1054 Unknown column 'k0_.number' in 'field list' at /opt/kimai/vendor/doctrine/dbal/src/Driver/PDO/Statement.php:130)"} {"channel":"request"}

```

Can someone help how I can fix this?

Running latest docker images from kimai/kimai2:fpm on a Synology nas with mysql 5.7.9 docker image!

**Comment by chrisonline** [comment]
Have fixe it via manual SQL Update who adds the new 2 columns, but would be good to find the cause.
Otherwise I have to manually upgrade again on the next version ;-)

**Reply by kevinpapst** [reply]
Nah, that's not the way... 
I assume you meant the activity/project columns, but manually trying to fix things is a bad idea, its impossible now to give any help.

The only migration that has `u1` is this here: https://github.com/kimai/kimai/blob/main/migrations/Version20231130000719.php
And there were two more migrations after that. 

Did you set the correct migration status?

**Reply by chrisonline** [reply]
Currently I have only executed this 2 SQL's from the 'Version20240326125247':
ALTER TABLE kimai2_activities ADD number VARCHAR(10) DEFAULT NULL'
ALTER TABLE kimai2_projects ADD number VARCHAR(10) DEFAULT NULL'

Now it is loading again.

But if you want I can of course drop this 2 columns so we can find the root cause.
Let me know.

**Reply by kevinpapst** [reply]
Try editing API password in your user profile, to see if you can create/read the new API tokens

**Reply by chrisonline** [reply]
Does not work. If I click on "API Zugang" I see an error 500.

Crash in the log:
```[2024-04-10T12:09:57.891812+00:00] request.CRITICAL: Uncaught PHP Exception Doctrine\DBAL\Exception\TableNotFoundException: "An exception occurred while executing a query: SQLSTATE[42S02]: Base table or view not found: 1146 Table 'kimai.kimai2_access_token' doesn't exist" at ExceptionConverter.php line 49 {"exception":"[object] (Doctrine\\DBAL\\Exception\\TableNotFoundException(code: 1146): An exception occurred while executing a query: SQLSTATE[42S02]: Base table or view not found: 1146 Table 'kimai.kimai2_access_token' doesn't exist at /opt/kimai/vendor/doctrine/dbal/src/Driver/API/MySQL/ExceptionConverter.php:49)\n[previous exception] [object] (Doctrine\\DBAL\\Driver\\PDO\\Exception(code: 1146): SQLSTATE[42S02]: Base table or view not found: 1146 Table 'kimai.kimai2_access_token' doesn't exist at /opt/kimai/vendor/doctrine/dbal/src/Driver/PDO/Exception.php:28)\n[previous exception] [object] (PDOException(code: 42S02): SQLSTATE[42S02]: Base table or view not found: 1146 Table 'kimai.kimai2_access_token' doesn't exist at /opt/kimai/vendor/doctrine/dbal/src/Driver/PDO/Statement.php:130)"} {"channel":"request"}```

**Reply by kevinpapst** [reply]
As expected. We should fix the SQL statement for you and then run the migrations.

**Comment by kevinpapst** [comment]
This seems to be the initial statement that is failing:
```
delete u1 from kimai2_user_preferences u1 where u1.name like '%.%' and exists(select 1 from (SELECT name, user_id FROM kimai2_user_preferences WHERE name LIKE '%_%') u2 where u2.name = replace(u1.name, '.', '_') and u1.user_id = u2.user_id)
```

I tested this on MySQL and MariaDB versions and it never failed on me, but maybe this is problematic for certain DB versions.
Can you try that manually in your docker

**Reply by chrisonline** [reply]
Done, this is the output:
![image](https://github.com/kimai/kimai/assets/2077218/37518d52-e36b-461e-a206-731ba2f0b410)

**Reply by kevinpapst** [reply]
That is annoying, so it is likely MySQL 5. With MariaDB 15 or MySQL 8 it works. 

Two things we could try:

1. What happens if you change it to:
```sql
delete from kimai2_user_preferences u1 where u1.name like '%.%' and exists(select 1 from (SELECT name, user_id FROM kimai2_user_preferences WHERE name LIKE '%_%') u2 where u2.name = replace(u1.name, '.', '_') and u1.user_id = u2.user_id)
```

and 2: If you have a backup, could you try to switch your docker compose to `mysql:8.3` and run it with that?

@tobybatch do you have a test system where we could verify that? Can I change all examples from 
```
  sqldb:
    image: mysql:5.7
```

to 
```
  sqldb:
    image: mysql:8.3
```

**Reply by chrisonline** [reply]
1. I get an error:
`ERROR 1064 (42000): You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near 'u1 where u1.name like '%.%' and exists(select 1 from (SELECT name, user_id FROM ' at line 1`

2. I will create a backup and a new installation. Report back after I am installed it.

**Reply by kevinpapst** [reply]
Thanks and Sorry 👍

**Reply by chrisonline** [reply]
I have tried to copy my current installation to use it as a test system, but if I use the mysql:**8.3** with my existing data I get some errors about "can't upgrade data from 5.x to 8.3".

So I can only install a new copy with 8.3.
But as far as you wrote with 8.3 it would work.

It is not better to have a test system with 5.x instead?

Sidenote:
On the documentation for "Docker compose" you have also listet 5.x and not 8.x!
So it would be good to change this that new users does not use the 5.x

https://www.kimai.org/documentation/docker-compose.html

**Reply by kevinpapst** [reply]
> On the documentation for "Docker compose" you have also listet 5.x and not 8.x!

I know:  https://github.com/kimai/www.kimai.org/pull/405

> It is not better to have a test system with 5.x instead?

Yes, I agree. Would be nice. But Kimai never was meant to run outdated software versions. 
And all of my managed systems work. 
On-Premise doesn't come with any kind of warranty.

Yes, it is a problem that docker compose docs recommend 5.3... but I hope we can change that soon.

> can't upgrade data from 5.x to 8.3

Thats fucked up. I am not a docker user, never was, never will, cant give support for that. 
So we have to wait if someone with more knowledge shows up.
We have to find a way to upgrade and then I will write a blog post about it.

**Reply by chrisonline** [reply]
> Thats fucked up. I am not a docker user, never was, never will, cant give support for that.
> So we have to wait if someone with more knowledge shows up.
> We have to find a way to upgrade and then I will write a blog post about it.

I understand. But it is not a docker problem.
It is a mysql 5.x upgrade to 8.x problem.
I will investigate this further and will try to find a way to upgrade.
Will let you know if I have success.

**Reply by chrisonline** [reply]
Success, I have now a kimai upgrade running with 8.3!
On 8.3 as you mentioned the upgrade has worked and the API settings are shown.

The main problem before was that I have not stopped the mysql 5.x before copying the data.

Now my upgrade path with docker was:
1.) Create docker with image **mysql 8**
2.) Stopped docker **mysql 5.x**
3.) Copied the **files** from **mysql 5.x** folder to **mysql 8.x** folder
4.) Started the **mysql 8** docker and the upgrade was made automatically

**Reply by kevinpapst** [reply]
Can you do me a favor and copy & paste the output of `mysql -v` ? I'd like to know the minor release version for the `DATABASE_URL`

```
       - DATABASE_URL=mysql://kimaiuser:kimaipassword@sqldb/kimai?charset=utf8mb4&serverVersion=8.3.0
```

**Reply by chrisonline** [reply]
Yes of course. Where should I call this command?
Inside the mysql docker container or kimai installation?

I have done it now in the mysql container:
```
bash-4.4# mysql -V 
mysql  Ver 8.3.0 for Linux on x86_64 (MySQL Community Server - GPL)
```

**Reply by kevinpapst** [reply]
Oh, hahaha, I didn't expect that it is really 8.3.**0** - thanks, then I can safely merge the documentation update

**Reply by kevinpapst** [reply]
And we'll see if others runs into the same issue. 
The update is out for two weeks now, so hopefully many people used mysql latest.

Thanks for helping with the tests!

**Reply by chrisonline** [reply]
> Thanks for helping with the tests!

Your welcome. 
I have now a Test system installed so if you need again help to test something out please let me know.

**Comment by nbaud** [comment]
I followed the steps in the documentation for upgrading, but stuck on this. I tried some manual tricks but to no avail. I am upgrading from `1.28.1` to `2.3.0` (I know it's a big step)

The main issue is this

```
[notice] Migrating up to DoctrineMigrations\Version20240926111739
[error] Migration DoctrineMigrations\Version20230327143628 failed during Execution. Error: "The table with name "kimai2.kimai2_working_times" already exists."
                                                                                                                        
 [ERROR] The table with name "kimai2.kimai2_working_times" already exists. 
```

from


```
Run composer recipes at any time to see the status of your Symfony recipes.

Executing script cache:clear [OK]
Executing script assets:install [OK]

nico@gigai1:/var/www/kimai2$ bin/console kimai:update

Kimai updates running ...
=========================

[notice] Migrating up to DoctrineMigrations\Version20240926111739
[error] Migration DoctrineMigrations\Version20230327143628 failed during Execution. Error: "The table with name "kimai2.kimai2_working_times" already exists."
                                                                                                                        
 [ERROR] The table with name "kimai2.kimai2_working_times" already exists.                                              
                                                                                                                        

nico@gigai1:/var/www/kimai2$ bin/console doctrine:migrations:status
+----------------------+----------------------+------------------------------------------------------------------------+
| Configuration                                                                                                        |
+----------------------+----------------------+------------------------------------------------------------------------+
| Storage              | Type                 | Doctrine\Migrations\Metadata\Storage\TableMetadataStorageConfiguration |
|                      | Table Name           | migration_versions                                                     |
|                      | Column Name          | version                                                                |
|----------------------------------------------------------------------------------------------------------------------|
| Database             | Driver               | Doctrine\DBAL\Driver\PDO\MySQL\Driver                                  |
|                      | Name                 | kimai2                                                                 |
|----------------------------------------------------------------------------------------------------------------------|
| Versions             | Previous             | DoctrineMigrations\Version20230126002049                               |
|                      | Current              | DoctrineMigrations\Version20230126002050                               |
|                      | Next                 | DoctrineMigrations\Version20230327143628                               |
|                      | Latest               | DoctrineMigrations\Version20240926111739                               |
|----------------------------------------------------------------------------------------------------------------------|
| Migrations           | Executed             | 62                                                                     |
|                      | Executed Unavailable | 0                                                                      |
|                      | Available            | 70                                                                     |
|                      | New                  | 8                                                                      |
|----------------------------------------------------------------------------------------------------------------------|
| Migration Namespaces | DoctrineMigrations   | /var/www/kimai2/migrations                                             |
+----------------------+----------------------+------------------------------------------------------------------------+
nico@gigai1:/var/www/kimai2$ 
```

**Reply by nbaud** [reply]
Also

```
nico@gigai1:/var/www/kimai2$ bin/console doctrine:migrations:status
+----------------------+----------------------+------------------------------------------------------------------------+
| Configuration                                                                                                        |
+----------------------+----------------------+------------------------------------------------------------------------+
| Storage              | Type                 | Doctrine\Migrations\Metadata\Storage\TableMetadataStorageConfiguration |
|                      | Table Name           | migration_versions                                                     |
|                      | Column Name          | version                                                                |
|----------------------------------------------------------------------------------------------------------------------|
| Database             | Driver               | Doctrine\DBAL\Driver\PDO\MySQL\Driver                                  |
|                      | Name                 | kimai2                                                                 |
|----------------------------------------------------------------------------------------------------------------------|
| Versions             | Previous             | DoctrineMigrations\Version20230126002049                               |
|                      | Current              | DoctrineMigrations\Version20230126002050                               |
|                      | Next                 | DoctrineMigrations\Version20230327143628                               |
|                      | Latest               | DoctrineMigrations\Version20240926111739                               |
|----------------------------------------------------------------------------------------------------------------------|
| Migrations           | Executed             | 62                                                                     |
|                      | Executed Unavailable | 0                                                                      |
|                      | Available            | 70                                                                     |
|                      | New                  | 8                                                                      |
|----------------------------------------------------------------------------------------------------------------------|
| Migration Namespaces | DoctrineMigrations   | /var/www/kimai2/migrations                                             |
+----------------------+----------------------+------------------------------------------------------------------------+

```

**Reply by nbaud** [reply]
that part

```
composer install --no-dev --optimize-autoloader
```

Went all fine

**Reply by kevinpapst** [reply]
What you describe doesn't make sense at all.
2.3.0 was released a year ago.
Also the table you mention is created in the middle of the entire process.
It can't already exist, unless your ran the upgrade before and it failed.
